File: core/src/main/java/com/lomekwi/cave/app/copy/CopyManager.java
package com.lomekwi.cave.app.copy;

import com.google.common.eventbus.Subscribe;
import com.lomekwi.cave.app.selection.SelectableSelectedEvent;
import org.jspecify.annotations.Nullable;

public class CopyManager {
    private @Nullable Copyable clipboard;
    private @Nullable Copyable latestCopyable;

    @Subscribe
    public void onSelection(SelectableSelectedEvent<?> e) {
        var sel = e.selectable();
        if (sel instanceof Copyable c) {
            latestCopyable = c;
        }
    }

    public void copy() {
        if (latestCopyable != null) {
            clipboard = latestCopyable.copy();
        }
    }

    public void copy(Copyable copyable) {
        clipboard = copyable;
    }

    public @Nullable Copyable getClipboard() {
        return clipboard;
    }

    public void clearClipboard() {
        clipboard = null;
    }

    public void refreshClipboard() {
        if (clipboard != null) {
            clipboard = clipboard.copy();
        }
    }
}

File: core/src/main/scala/com/lomekwi/cave/app/copy/CopyManager.scala
package com.lomekwi.cave.app.copy

import com.google.common.eventbus.Subscribe
import com.lomekwi.cave.app.selection.SegmentSetSelectedEvent

import scala.compiletime.uninitialized

class CopyManager {
  private var clipboard: Copyable = uninitialized
  private var latestCopyable: Copyable = uninitialized

  /** 复制以"当前选中集"为单位，单选也是只有一项的集合。 */
  @Subscribe
  def onSelection(e: SegmentSetSelectedEvent): Unit = {
    latestCopyable = e.set
  }

  @Subscribe
  def onCopyEvent(e: CopyEvent): Unit = {
    e.target match {
      case c: Copyable => clipboard = c.copy()
      case _ =>
    }
  }

  def copy(): Unit = {
    if (latestCopyable != null) {
      clipboard = latestCopyable.copy()
    }
  }

  def copy(copyable: Copyable): Unit = {
    clipboard = copyable
  }

  def getClipboard: Copyable = {
    clipboard
  }

  def clearClipboard(): Unit = {
    clipboard = null
  }

  def refreshClipboard(): Unit = {
    if (clipboard != null) {
      clipboard = clipboard.copy()
    }
  }
}
